### PR TITLE
Reentrancy counter agent

### PR DESCRIPTION
**Pull request by e1Ru1o**
Agent that detect reentracy calls using the call stack provided in the tx traces

### CODE REVIEW

File: reentrancy-counter/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# High Gas Agent
+
```
Comment (rachit2501) [inline reentrancy-counter/README.md:2]: Readme needs to be updated

File: reentrancy-counter/src/agent.spec.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        createTrace("0x3", [0, 0, 0]),
+      ]);
+      const findings: Finding[] = await handleTransaction(tx);
+      expect(findings).toStrictEqual([]);
+    });
+
+    it("Should ignore non reentrant calls", async () => {
+      const tx: TransactionEvent = createTxEvent([
+        createTrace("0x0", []),
+        createTrace("0x1", [0]),
```
Comment (rachit2501) [inline reentrancy-counter/src/agent.spec.ts:57]: More description on what 0, 1 ... represents ( array elements ) would be nice to have in comments

File: reentrancy-counter/src/agent.spec.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        createTrace("0x2", [0, 0]),
+        createTrace("0x3", [0, 0, 0]),
+        createTrace("0x2", [0, 0, 0, 0]),
+        createTrace("0x3", [0, 0, 0, 0, 0]),
+        createTrace("0x2", [0, 0, 0, 0, 0, 0]),
+        createTrace("0x4", [0, 1]),
+        createTrace("0x5", [0, 1, 0]),
+        createTrace("0x6", [0, 1, 0, 0]),
+        createTrace("0x4", [0, 1, 0, 0, 0]),
```
Comment (rachit2501) [inline reentrancy-counter/src/agent.spec.ts:89]: can you add a small diag showing the loops/flow, a link to some diag would be great.

**Review by rachit2501 — CHANGES_REQUESTED** [review]

File: reentrancy-counter/src/agent.spec.ts (hunk 2 above)
Comment (e1Ru1o) [reply reentrancy-counter/src/agent.spec.ts:57]: as you can see in the `createTrace` function the array represent the `traceAddress` field of a `Trace` that is why no explanation was added there, but in short the array represent the sequence of calls made that originated the `Trace` and the number represent the depth of each call.
You can read about it here https://openethereum.github.io/JSONRPC-trace-module

File: reentrancy-counter/src/agent.spec.ts (hunk 3 above)
Comment (e1Ru1o) [reply reentrancy-counter/src/agent.spec.ts:89]: Check the link above an tell me if the diag is still needed.

Comment (rachit2501) [reply reentrancy-counter/src/agent.spec.ts:89]: Yes. I read the link. Would be better if you can provide a trace similar to in the example.

Comment (rachit2501) [reply reentrancy-counter/src/agent.spec.ts:89]: A
  CALLs B
    CALLs G
  CALLs C
    CALLs G

Comment (e1Ru1o) [reply reentrancy-counter/src/agent.spec.ts:89]: Well, I think that the calls all self explanatory (the stack is been showed), but I am going to add some indentation in the calls in order to simulate the approach.

Comment (e1Ru1o) [reply reentrancy-counter/src/agent.spec.ts]: The indentation looks very bad, then some comments added int the test (note that calls have the same shape of the stacks). I hope it looks clearer for you now.

**Review by rachit2501 — APPROVED** [review]

**Pull request merged**